File: part-1/src/main/java/conc/view/SimulationView.java
package conc.view;



import conc.model.Body;
import conc.model.Boundary;
import conc.model.InputListener;
import conc.model.P2d;

import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.util.ArrayList;
import java.util.List;

/**
 * Simulation view
 *
 * @author aricci
 *
 */
public class SimulationView {
        
	private final VisualiserFrame frame;
	
    /**
     * Creates a view of the specified size (in pixels)
     * 
     * @param w width.
     * @param h height.
     */
    public SimulationView(int w, int h){
    	frame = new VisualiserFrame(w,h);
    }
        
    public void display(List<Body> bodies, double vt, long iter, Boundary bounds){
 	   frame.display(bodies, vt, iter, bounds); 
    }

	public void registerListener(InputListener l){
		frame.registerListener(l);
	}
    
    public static class VisualiserFrame extends JFrame implements ActionListener {
        private final VisualiserPanel panel;
		private final List<InputListener> listeners = new ArrayList<>();

		public VisualiserFrame(int w, int h){
            setTitle("Bodies Simulation");
            setSize(w+100,h+100);
            setResizable(false);

            panel = new VisualiserPanel(w,h);
			JPanel controlPanel = new JPanel();
			JButton start = new JButton("start");
			controlPanel.add(start);
			JButton stop = new JButton("stop");
			controlPanel.add(stop);
			LayoutManager layout = new BorderLayout();
			JPanel cp = new JPanel();
			cp.setLayout(layout);
			cp.add(BorderLayout.NORTH, controlPanel);
			cp.add(BorderLayout.CENTER, panel);
			setContentPane(cp);

			start.addActionListener(this);
			stop.addActionListener(this);
            addWindowListener(new WindowAdapter(){
    			public void windowClosing(WindowEvent ev){
    				System.exit(-1);
    			}
    			public void windowClosed(WindowEvent ev){
    				System.exit(-1);
    			}
    		});
    		this.setVisible(true);
        }
        
        public void display(List<Body> bodies, double vt, long iter, Boundary bounds){
        	try {
	        	SwingUtilities.invokeLater(() -> {
	        		panel.display(bodies, vt, iter, bounds);
	            	repaint();
	        	});
        	} catch (Exception ex) {ex.printStackTrace();}
        }

        public void updateScale(double k) {
        	panel.updateScale(k);
        }

		@Override
		public void actionPerformed(ActionEvent e) {
			String cmd = e.getActionCommand();
			if (cmd.equals("start")){
				notifyStarted();
			} else if (cmd.equals("stop")){
				notifyStopped();
			}
		}

		public void registerListener(InputListener listener){
			this.listeners.add(listener);
		}

		public void notifyStarted(){
			listeners.forEach(InputListener::started);
		}

		public void notifyStopped(){
			listeners.forEach(InputListener::stopped);
		}
    }

    public static class VisualiserPanel extends JPanel implements KeyListener {
        
    	private List<Body> bodies;
    	private Boundary bounds;
    	
    	private long nIter;
    	private double vt;
    	private double scale = 1;
    	
        private final long dx;
        private final long dy;
        
        public VisualiserPanel(int w, int h){
            setSize(w,h);
            dx = w/2 - 20;
            dy = h/2 - 20;
			this.addKeyListener(this);
			setFocusable(true);
			setFocusTraversalKeysEnabled(false);
			requestFocusInWindow(); 
        }

        public void paint(Graphics g){    		    		
    		if (bodies != null) {
        		Graphics2D g2 = (Graphics2D) g;
        		
        		g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING,
        		          RenderingHints.VALUE_ANTIALIAS_ON);
        		g2.setRenderingHint(RenderingHints.KEY_RENDERING,
        		          RenderingHints.VALUE_RENDER_QUALITY);
        		g2.clearRect(0,0,this.getWidth(),this.getHeight());

        		
        		int x0 = getXcoord(bounds.getX0());
        		int y0 = getYcoord(bounds.getY0());
        		
        		int wd = getXcoord(bounds.getX1()) - x0;
        		int ht = y0 - getYcoord(bounds.getY1());
        		
    			g2.drawRect(x0, y0 - ht, wd, ht);
    			
	    		bodies.forEach( b -> {
	    			P2d p = b.getPos();
			        int radius = (int) (10*scale);
			        if (radius < 1) {
			        	radius = 1;
			        }
			        g2.drawOval(getXcoord(p.getX()),getYcoord(p.getY()), radius, radius); 
			    });		    
	    		String time = String.format("%.2f", vt);
	    		g2.drawString("Bodies: " + bodies.size() + " - vt: " + time + " - nIter: " + nIter + " (UP for zoom in, DOWN for zoom out)", 2, 20);
    		}
        }
        
        private int getXcoord(double x) {
        	return (int)(dx + x*dx*scale);
        }

        private int getYcoord(double y) {
        	return (int)(dy - y*dy*scale);
        }
        
        public void display(List<Body> bodies, double vt, long iter, Boundary bounds){
            this.bodies = bodies;
            this.bounds = bounds;
            this.vt = vt;
            this.nIter = iter;
        }
        
        public void updateScale(double k) {
        	scale *= k;
        }

		@Override
		public void keyPressed(KeyEvent e) {
				if (e.getKeyCode() == 38){  		/* KEY UP */
					scale *= 1.1;
				} else if (e.getKeyCode() == 40){  	/* KEY DOWN */
					scale *= 0.9;  
				} 
		}

		public void keyReleased(KeyEvent e) {}
		public void keyTyped(KeyEvent e) {}
	}
}

File: part-1/src/main/scala/view/SimulationView.scala
package view

import view.SimulationView
import view.ViewUtils
import com.sun.java.accessibility.util.AWTEventMonitor.{addKeyListener, addWindowListener}
import model.Body
import model.Boundary
import model.Objects2d.P2d

import java.awt.event.{ActionListener, KeyEvent, KeyListener, WindowAdapter, WindowEvent}
import java.awt.{BorderLayout, Dimension, Graphics2D, RenderingHints}
import javax.swing.SwingUtilities
import scala.language.postfixOps
import scala.swing.*
import scala.swing.BorderPanel.Position.*
import scala.swing.event.ButtonClicked

class SimulationView(view: ViewUtils, w: Int, h: Int) extends Frame:
  val visualiserPanel = new VisualiserPanel(w,h, view.bounds)
  size = Dimension(w + 100, h + 100)
  title = "Bodies simulation"
  resizable = false
  contents = new BorderPanel{
    layout(ControlPanel(view)) = North
    layout(visualiserPanel) = Center
  }
  visible = true
  addWindowListener( new WindowAdapter {
    override def windowClosing(ev: WindowEvent): Unit = System.exit(-1)
    override def windowClosed(ev: WindowEvent): Unit = System.exit(-1)
  })
  def display(bodies: List[Body], vt: Double, iter: Long, bounds: Boundary): Unit = //todo add here actor?
    SwingUtilities.invokeLater(() =>
      visualiserPanel display(bodies, vt, iter)
      repaint()
    )
end SimulationView

sealed class ControlPanel(view: ViewUtils) extends FlowPanel:
  val start: Button = new Button("start"){
    reactions += {
      case event.ButtonClicked(_) =>
        enabled = false
        stop.enabled = true
        view.startSimulation()
    }
  }
  val stop: Button = new Button("stop"){
    reactions += {
      case event.ButtonClicked(_) =>
        enabled = false
        start.enabled = true
        view.stopSimulation()
    }
  }
  contents += start
  contents += stop
end ControlPanel


sealed class VisualiserPanel(w: Int, h: Int, bound: Boundary) extends Panel, KeyListener:
  var bodies: List[Body] = List()
  var nIter: Int = 0
  var vt: Double = 0
  var scale: Double = 1
  val dx: Long = w/2 -20
  val dy: Long = h/2 -20

  preferredSize = Dimension(w,h)

  override def paint(g: Graphics2D): Unit =
    if bodies.nonEmpty then
      val g2: Graphics2D = g
      g2 setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON)
      g2 setRenderingHint(RenderingHints.KEY_RENDERING, RenderingHints.VALUE_RENDER_QUALITY)
      g2 clearRect(0, 0, w, h)

      val x0 = getXCoord(bound.x0)
      val y0 = getYCoord(bound.y0)
      val wd = getXCoord(bound.x1) - x0
      val ht = y0 - getYCoord(bound.y1)

      g2 drawRect(x0, y0 - ht, wd, ht)

      bodies.foreach( b =>
        val p: P2d = b.pos
        var radius: Int = (10 * scale).toInt
        if radius < 1 then radius = 1
        g2 drawOval(getXCoord(p.x), getYCoord(p.y), radius, radius)
      )
      val time: String = String format("%.2f", vt)
      g2 drawString("Bodies: " + bodies.length + " - vt: " + time + " - nIter: " + nIter + " (UP for zoom in, DOWN for zoom out)", 2, 20)

  def getXCoord(x: Double): Int = (dx + (x * dx * scale)).toInt
  def getYCoord(y: Double): Int = (dy - (y * dy * scale)).toInt

  def display(b: List[Body], v: Double, iter: Long): Unit =
    this.bodies = b
    this.vt = v
    this.nIter = iter.toInt

  override def keyPressed(e: KeyEvent): Unit = e.getKeyCode match
    case KeyEvent.VK_UP => updateScale(1.1)
    case KeyEvent.VK_DOWN => updateScale(0.9)

  override def keyTyped(e: KeyEvent): Unit = {}
  override def keyReleased(e: KeyEvent): Unit = {}
  private def updateScale(k: Double): Unit = scale *= k

end VisualiserPanel
